book_maker/translator/chatgptapi_translator.py

```diff
@@ -90,5 +122,128 @@ def translate(self, text):
             t_text = self.get_translation(text)
 
         # todo: Determine whether to print according to the cli option
-        print(t_text.strip())
+        if needprint:
+            print(re.sub("\n{3,}", "\n\n", t_text))
+
+        elapsed_time = time.time() - start_time
+        print(f"translation time: {elapsed_time:.1f}s")
+
         return t_text
+
+    def translate_and_split_lines(self, text):
+        result_str = self.translate(text, False)
+        lines = result_str.split("\n")
+        lines = [line.strip() for line in lines if line.strip() != ""]
+        return lines
+
+    def get_best_result_list(
+        self, plist_len, new_str, sleep_dur, result_list, max_retries=15
+    ):
+        if len(result_list) == plist_len:
+            return result_list, 0
+
+        best_result_list = result_list
+        retry_count = 0
+
+        while retry_count < max_retries:
+            print(
+                f"bug: {plist_len} -> {len(result_list)} : Number of paragraphs before and after translation"
+            )
+            print(f"sleep for {sleep_dur}s and retry {retry_count+1} ...")
+            time.sleep(sleep_dur)
+            retry_count += 1
+            result_list = self.translate_and_split_lines(new_str)
+            if (
+                len(result_list) == plist_len
+                or len(best_result_list) < len(result_list) <= plist_len
+                or (
+                    len(result_list) < len(best_result_list)
+                    and len(best_result_list) > plist_len
+                )
+            ):
+                best_result_list = result_list
+
+            if len(result_list) == plist_len:
+                break
+
+        return best_result_list, retry_count
+
+    def log_retry(self, state, retry_count, elapsed_time, log_path="log/buglog.txt"):
```

Comment:
Previously, I forgot about this. We need to ensure that the user has the 'log' directory. If it does not exist, create this; otherwise, the `write` operation will fail. You can refer to https://stackoverflow.com/questions/23793987/write-file-to-a-directory-that-doesnt-exist.

Reply:
I added 
```
if not os.path.exists("log"):
   os.makedirs("log")
```
in epub_loader because now only when translate epub and use `--accumulated_num` will write to here

Reply:
Got it